### PR TITLE
[mlir][linalg] Add Grouped Convolution Ops: conv_2d_nhwgc_gfhwc and conv_2d_nhwgc_gfhwc_q

### PR DESCRIPTION
**Pull request by stefankoncarevic**
This patch adds two new ops: linalg::Conv2DNhwgcGfhwcOp and linalg::Conv2DNhwgcGfhwcQOp, and uses them to convert tosa group conv2d Ops.
- Added linalg::Conv2DNhwgcGfhwcOp and linalg::Conv2DNhwgcGfhwcQOp.
- Updated the conversion process to use these new ops for tosa group conv2d operations.

### CODE REVIEW

**Comment by sjarus** [comment]
Hi there, sorry for the delay and thanks for this contribution! As @GeorgeARM  mentioned, TOSA has a [specification](https://www.mlplatform.org/tosa/tosa_spec.html) that the dialect implements. This is because the spec is also used to design hardware and backend software stacks that depend on the existence of a stable IR construct. 

We'd be happy to help you/your organization contribute to the specification with this proposal and it'll then work its way into the dialect. However, this will take more time to add, than the time to make a dialect change that has no specification impact (e.g. update a verifier) 

Until then, to unblock you is it possible to address this as two separate things - the change to the dialect (with its spec relationship) and the TosaToLinalg pass. Is it an option to refactor the latter to pattern match multiple tosa.conv2d instances that constitute a grouped convolution ?

**Review by eric-k256 — DISMISSED** [review]
+1 on what @sjarus said. This would be a notable deviation from the TOSA spec. 

Marking this as requesting changes to note the request to split the linalg change from the TOSA change.

**Comment by stefankoncarevic** [comment]
Hi everyone,  
I've made the requested changes by separating the Linalg grouped convolution operations into this PR. This will allow us to proceed with upstreaming just this part.  
Regarding the TOSA dialect changes, after discussions within our team, we've decided not to upstream that portion as it diverges too much from the current specification. We'll focus on the Linalg operations for now.  
Thank you for your feedback and support!

**Comment by eric-k256** [comment]
Thanks for the update. We're always looking for feedback and future changes we should be making to TOSA, feel free to reach out for a discussion on the topic.
I've removed my change request, one of the other reviewers would be better placed to approve the change.

**Comment by krzysz00** [comment]
Ping for people who know `linalg`

**Comment by bjacob** [comment]
FYI @rsuderman, @MaheshRavishankar, @hanhanW.

**Review by MaheshRavishankar — APPROVED** [review]

File: mlir/test/Dialect/Linalg/named-ops.mlir [hunk 1]
```diff
@@ -409,6 +409,22 @@ func.func @conv_2d_ngchw_fgchw(%input: memref<?x?x?x?x?xf32>, %filter: memref<?x
 
 // -----
 
+// CHECK-LABEL: func @conv_2d_nhwgc_gfhwc
```
Comment (hanhanW) [inline mlir/test/Dialect/Linalg/named-ops.mlir:412]: Can you add a test case for static shapes, like below test case? It is not easy to see if the indexing maps are defined properly or not. Having a static shape test case could capture the failure if there is something wrong.

**Review by hanhanW — APPROVED** [review]

**Comment by stefankoncarevic** [comment]
Hi @krzysz00, @manupak can you merge PR, I don't have rights.

**Comment by rengolin** [comment]
Just a comment, this has the same problems as #113953.

I'm not sure what the difference is here and why people are agreeing with this one and not the other one.

**Review by rengolin — CHANGES_REQUESTED** [review]
Shouldn't we talk about the more general pattern here instead of adding dozens of variations on the same theme?

**Comment by stellaraccident** [comment]
> Shouldn't we talk about the more general pattern here instead of adding dozens of variations on the same theme?

Yes, I'm having trouble with the firehose of individual conv op additions, and responding to individual threads isn't working. Let's boost this to the forum and list the component patches to discuss together. 

FTR - these seem to be very basic variations of conv needed to support common use cases and I am sympathetic to the need to have a practical solution for them in short order. I'm also cringing at the addition of technical debt this adds because the project overall doesn't have a plan for how to manage the expansion. By raising this to the forum, hopefully we can quickly converge on a practical path to land these now and start the conversation on the ultimate plan. We just did that for matmul and it wasn't too bad -- so I'm hopeful about these too.

**Comment by rengolin** [comment]
> By raising this to the forum, hopefully we can quickly converge on a practical path to land these now and start the conversation on the ultimate plan. We just did that for matmul and it wasn't too bad -- so I'm hopeful about these too.

https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863

**Review by rengolin — APPROVED** [review]
Removing the block so that we have the discussion in the forums. But let's not add more stuff, please.

**Pull request merged**